src/main/scala/com/fulcrumgenomics/commons/util/Async.scala

```diff
@@ -0,0 +1,286 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.commons.util
+
+import java.io.Closeable
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger, AtomicReference}
+import java.util.concurrent.{ArrayBlockingQueue, BlockingQueue, TimeUnit}
+
+import com.fulcrumgenomics.commons.io.Writer
+
+import scala.concurrent.forkjoin.ForkJoinPool
+
+/**
+  * Provides generic utility classes related to asynchronous processing, reading, writing, and iterating over items.
+  */
+object Async {
+
+  /** The default # of items to buffer in queues. */
+  private val DefaultBufferSize: Int = 2048
+
+  private object AsyncIterator {
+    /** A [[Runnable]] who can be asked when it is all done.
+      *
+      * @param f the method to run
+      */
+    case class Runner(f: () => Unit) extends Runnable {
+      private val allDone = new AtomicBoolean(false)
+
+      def run(): Unit = {
+        f()
+        this.allDone.set(true)
+      }
+
+      def done: Boolean = this.allDone.get()
+    }
+  }
+
+  /** An asynchronous wrapper for an [[Iterator]].  A separate thread will be created to consume the source iterator.
+    * Will buffer up to [[bufferSize]] elements before the blocking when consuming the source.
+    *
+    * @param source the source iterator to consume
+    * @param bufferSize the maximum number of elements to buffer before the blocking when consuming the source
+    * @tparam T the type of object to consume
+    */
+  class AsyncIterator[T](val source: Iterator[T], bufferSize: Int = DefaultBufferSize) extends Iterator[T] {
+    require(bufferSize > 0, "bufferSize must be greater than zero")
+    private val queue: BlockingQueue[T] = new ArrayBlockingQueue[T](bufferSize)
+    private val outer = this
+    private val runnable = AsyncIterator.Runner(() => outer.source.foreach(outer.queue.put))
+    private val thread = {
+      val t = new Thread(runnable, getClass.getSimpleName)
+      t.setDaemon(true)
+      t.start()
+      t
+    }
+
+    /** Returns true if there exists more elements, false otherwise */
+    def hasNext(): Boolean = {
+      if (!this.queue.isEmpty || this.source.nonEmpty) true
```

Comment:
I don't think you should ever touch `source` outside of `runnable`.  What if `nonEmpty` is side-effecting, e.g. it tries to queue up the next item?  At the very least you need to add synchronization so that here and in the runnable you can't be hitting source at the same time, but I'm not sure that's sufficient.